sw/device/silicon_creator/lib/drivers/flash_ctrl.c

```diff
@@ -23,99 +35,121 @@ static bool is_busy(void) {
-                "Incorrect enum value for kFlashCtrlRegionData");
-  static_assert(
-      kFlashCtrlRegionInfo0 == 1u << FLASH_CTRL_CONTROL_PARTITION_SEL_BIT,
-      "Incorrect enum value for kFlashCtrlRegionInfo0");
-  static_assert(
-      kFlashCtrlRegionInfo1 == (1u << FLASH_CTRL_CONTROL_PARTITION_SEL_BIT |
-                                1u << FLASH_CTRL_CONTROL_INFO_SEL_OFFSET),
-      "Incorrect enum value for kFlashCtrlRegionInfo1");
-  static_assert(
-      kFlashCtrlRegionInfo2 == (1u << FLASH_CTRL_CONTROL_PARTITION_SEL_BIT |
-                                2u << FLASH_CTRL_CONTROL_INFO_SEL_OFFSET),
-      "Incorrect enum value for kFlashCtrlRegionInfo2");
-  static_assert(kFlashCtrlErasePage == 0u,
-                "Incorrect enum value for kFlashCtrlErasePage");
-  static_assert(kFlashCtrlEraseBank == 1u << FLASH_CTRL_CONTROL_ERASE_SEL_BIT,
-                "Incorrect enum value for kFlashCtrlEraseBank");
-
-  // Set the partition.
-  control_reg_val |= (uint32_t)region;
-
-  // Set the erase type.
-  control_reg_val |= (uint32_t)erase_type;
-
-  // Set the number of words as `word_count - 1` as noted in #3353.
-  control_reg_val = bitfield_field32_write(
-      control_reg_val, FLASH_CTRL_CONTROL_NUM_FIELD, word_count - 1);
-
-  // Start the transaction.
-  control_reg_val =
-      bitfield_bit32_write(control_reg_val, FLASH_CTRL_CONTROL_START_BIT, true);
-
-  // Write the configuration.
-  abs_mmio_write32(kBase + FLASH_CTRL_CONTROL_REG_OFFSET, control_reg_val);
-
+  abs_mmio_write32(kBase + FLASH_CTRL_ADDR_REG_OFFSET, params.addr);
+  // Configure flash_ctrl and start the transaction.
+  uint32_t reg = bitfield_bit32_write(0, FLASH_CTRL_CONTROL_START_BIT, true);
+  reg =
+      bitfield_field32_write(reg, FLASH_CTRL_CONTROL_OP_FIELD, params.op_type);
+  reg = bitfield_bit32_write(reg, FLASH_CTRL_CONTROL_PARTITION_SEL_BIT,
+                             params.partition != kFlashCtrlPartitionData);
+  reg = bitfield_field32_write(reg, FLASH_CTRL_CONTROL_INFO_SEL_FIELD,
+                               (uint32_t)params.partition >> 1);
+  reg = bitfield_bit32_write(reg, FLASH_CTRL_CONTROL_ERASE_SEL_BIT,
+                             params.erase_type == kFlashCtrlEraseTypeBank);
+  // TODO(#3353): Remove -1 when flash_ctrl is updated.
+  reg = bitfield_field32_write(reg, FLASH_CTRL_CONTROL_NUM_FIELD,
+                               params.word_count - 1);
+  abs_mmio_write32(kBase + FLASH_CTRL_CONTROL_REG_OFFSET, reg);
   return kErrorOk;
 }
 
+/**
+ * Copies data from the flash_ctrl FIFO to the given buffer.
+ *
+ * Large reads may create back pressure.
+ *
+ * @param word_count Number of words to read from the FIFO.
+ * @param[out] data_out Output buffer.
+ */
 static void fifo_read(size_t word_count, uint32_t *data_out) {
-  // Keep reading until no words remain. For large reads this may create back
-  // pressure.
   for (size_t words_read = 0; words_read < word_count; ++words_read) {
```

Comment:
optional: I'd probably `s/words_read/i/` for consistency with other code while we are here.

Same for `words_programmed` below.

Reply:
I agree. Also renamed fifo_push to fifo_write for consistency.